File: solutions_2453486_1/java/cjoa/a_tic_tac_toe_tomek.java
import java.io.*;

public class a_tic_tac_toe_tomek {
   
   public static void main (String[] args) throws IOException {
      (new a_tic_tac_toe_tomek()).Run();
   }

   boolean won(char player, char[][] B) {
      boolean ok;
      for (int r = 0; r < 4; ++r) {
         ok = true;
         for (int c = 0; c < 4; ++c) {
            if (B[r][c] != player && B[r][c] != 'T') {
               ok = false;
               break;
            }
         }
         if (ok) return true;
      }

      for (int c = 0; c < 4; ++c) {
         ok = true;
         for (int r = 0; r < 4; ++r) {
            if (B[r][c] != player && B[r][c] != 'T') {
               ok = false;
               break;
            }
         }
         if (ok) return true;
      }

      ok = true;
      for (int i = 0; i < 4; ++i) {
         if (B[i][i] != player && B[i][i] != 'T') {
            ok = false;
            break;
         }
      }
      if (ok) return true;

      ok = true;
      for (int i = 0; i < 4; ++i) {
         if (B[i][3-i] != player && B[i][3-i] != 'T') {
            ok = false;
            break;
         }
      }
      if (ok) return true;

      return false;
   }
   
   int countPlays(char[][] B) {
      int res = 0;
      for (int r = 0; r < 4; ++r)
         for (int c = 0; c < 4; ++c)
            if (B[r][c] != '.')
               ++res;
      return res;
   }
   
   String getState(char[][] B) {
      if (won('X', B)) return "X won";
      if (won('O', B)) return "O won";
      int cnt = countPlays(B);
      return cnt == 16 ? "Draw" : "Game has not completed";
   }
   
   void Run() throws IOException {
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
      int TC = Integer.parseInt( in.readLine().trim() );
      char[][] B = new char[4][];
      for (int tc = 1; tc <= TC; ++tc) {
         for (int r = 0; r < 4; ++r)
            B[r] = in.readLine().trim().toCharArray();
         in.readLine();
         System.out.println("Case #" + tc + ": " + getState(B));
      }
   }
}

File: solutions_2453486_0/Python/cjoa/a_tic_tac_toe_tomek.py
#!/usr/bin/env python

from sys import stdin

def game_over(B):
   return all( [ B[r][c] != '.' for r in xrange(4) for c in xrange(4) ] )

def is_winner(player, B):
   for r in xrange(4):
      if all( [ B[r][c] == player or B[r][c] == 'T' for c in xrange(4) ] ):
         return True
   for c in xrange(4):
      if all( [ B[r][c] == player or B[r][c] == 'T' for r in xrange(4) ] ):
         return True
   if all( [ B[i][i] == player or B[i][i] == 'T' for i in xrange(4) ] ):
      return True
   if all( [ B[i][3-i] == player or B[i][3-i] == 'T' for i in xrange(4) ] ):
      return True
   return False

def get_state(B):
   if is_winner('X', B): return "X won"
   if is_winner('O', B): return "O won"
   return "Draw" if game_over(B) else "Game has not completed"

def main():
   TC = int(stdin.readline().strip())
   for tc in xrange(1, TC+1):
      B = []
      for r in xrange(4):
         B.append( stdin.readline().strip() )
      stdin.readline()
      print "Case #%d: %s" % (tc, get_state(B))
   return 0

if __name__ == '__main__': main()
